source/fuzz/available_instructions.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021 Alastair F. Donaldson
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/available_instructions.h"
+
+namespace spvtools {
+namespace fuzz {
+
+AvailableInstructions::AvailableInstructions(
+    opt::IRContext* ir_context,
+    const std::function<bool(opt::IRContext*, opt::Instruction*)>& filter)
+    : ir_context_(ir_context) {
+  // Consider all global declarations
+  for (auto& global : ir_context->module()->types_values()) {
+    if (filter(ir_context, &global)) {
+      available_globals_.push_back(&global);
+    }
+  }
+
+  for (auto& function : *ir_context->module()) {
+    std::vector<opt::Instruction*> params;
+    function.ForEachParam(
+        [this, &filter, ir_context, &params](opt::Instruction* param) {
+          if (filter(ir_context, param)) {
+            params.push_back(param);
+          }
+        });
+    available_params_.insert({&function, params});
```

Comment:
We could probably use `available_params_.emplace(&function, std::move(params))`.

Reply:
Since you are using `available_params_` below it might be better to move this to the end of the loop or store its size in a variable.

Reply:
Done.